File: internal/email/hwm.go
package email

import (
	"context"
	"fmt"
	"sync"
)

type stagedHWMRecord struct {
	host        string
	folder      string
	uidValidity uint32
	lastUID     uint32
}

// StagingHWMStore keeps HWM updates in memory until Commit. It is used by
// InboxService to make reply persistence happen before durable HWM advancement.
type StagingHWMStore struct {
	underlying HWMStore
	mu         sync.Mutex
	staged     []stagedHWMRecord
}

func NewStagingHWMStore(underlying HWMStore) *StagingHWMStore {
	return &StagingHWMStore{underlying: underlying}
}

func (s *StagingHWMStore) Get(ctx context.Context, host, folder string) (*uint32, *uint32, error) {
	s.mu.Lock()
	for i := len(s.staged) - 1; i >= 0; i-- {
		if s.staged[i].host == host && s.staged[i].folder == folder {
			v, u := s.staged[i].uidValidity, s.staged[i].lastUID
			s.mu.Unlock()
			return &v, &u, nil
		}
	}
	underlying := s.underlying
	s.mu.Unlock()
	if underlying == nil {
		return nil, nil, nil
	}
	return underlying.Get(ctx, host, folder)
}

func (s *StagingHWMStore) Set(_ context.Context, host, folder string, uidValidity, lastUID uint32) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.staged = append(s.staged, stagedHWMRecord{host: host, folder: folder, uidValidity: uidValidity, lastUID: lastUID})
	return nil
}

func (s *StagingHWMStore) Commit(ctx context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.underlying == nil {
		s.staged = nil
		return nil
	}
	for _, rec := range s.staged {
		if err := s.underlying.Set(ctx, rec.host, rec.folder, rec.uidValidity, rec.lastUID); err != nil {
			return fmt.Errorf("email: commit high-water mark (%s/%s): %w", rec.host, rec.folder, err)
		}
	}
	s.staged = nil
	return nil
}

File: crates/symeraseme-core/src/email/hwm.rs
//! High-water-mark stores. The staging store is what makes reply persistence
//! happen before a durable HWM advance: a failed insert leaves the persisted
//! mark untouched, so a retry can ingest the same batch again.

use std::collections::HashMap;
use std::sync::Mutex;

/// Persists the last processed UID per host/folder together with the
/// UIDVALIDITY it was observed under. A UIDVALIDITY change forces a cold start.
pub trait HwmStore: Send + Sync {
    fn get(&self, host: &str, folder: &str) -> Result<(Option<u32>, Option<u32>), String>;
    fn set(&self, host: &str, folder: &str, uid_validity: u32, last_uid: u32)
    -> Result<(), String>;
}

fn key(host: &str, folder: &str) -> String {
    format!("{host}\u{0}{folder}")
}

#[derive(Default)]
pub struct MemoryHwmStore {
    records: Mutex<HashMap<String, (u32, u32)>>,
}

impl MemoryHwmStore {
    pub fn new() -> Self {
        Self::default()
    }
}

impl HwmStore for MemoryHwmStore {
    fn get(&self, host: &str, folder: &str) -> Result<(Option<u32>, Option<u32>), String> {
        let records = self.records.lock().expect("hwm store lock");
        Ok(match records.get(&key(host, folder)) {
            Some((validity, last)) => (Some(*validity), Some(*last)),
            None => (None, None),
        })
    }

    fn set(
        &self,
        host: &str,
        folder: &str,
        uid_validity: u32,
        last_uid: u32,
    ) -> Result<(), String> {
        let mut records = self.records.lock().expect("hwm store lock");
        records.insert(key(host, folder), (uid_validity, last_uid));
        Ok(())
    }
}

/// Keeps HWM updates in memory until [`StagingHwmStore::commit`].
pub struct StagingHwmStore<'a> {
    underlying: Option<&'a dyn HwmStore>,
    staged: Mutex<Vec<(String, String, u32, u32)>>,
}

impl<'a> StagingHwmStore<'a> {
    pub fn new(underlying: Option<&'a dyn HwmStore>) -> Self {
        Self {
            underlying,
            staged: Mutex::new(Vec::new()),
        }
    }

    /// Writes every staged record through, in the order it was staged. The
    /// first failure stops the commit and keeps the remaining records staged.
    pub fn commit(&self) -> Result<(), String> {
        let mut staged = self.staged.lock().expect("hwm store lock");
        let Some(underlying) = self.underlying else {
            staged.clear();
            return Ok(());
        };
        for (host, folder, uid_validity, last_uid) in staged.iter() {
            underlying
                .set(host, folder, *uid_validity, *last_uid)
                .map_err(|error| {
                    format!("email: commit high-water mark ({host}/{folder}): {error}")
                })?;
        }
        staged.clear();
        Ok(())
    }
}

impl HwmStore for StagingHwmStore<'_> {
    fn get(&self, host: &str, folder: &str) -> Result<(Option<u32>, Option<u32>), String> {
        {
            let staged = self.staged.lock().expect("hwm store lock");
            for (staged_host, staged_folder, uid_validity, last_uid) in staged.iter().rev() {
                if staged_host == host && staged_folder == folder {
                    return Ok((Some(*uid_validity), Some(*last_uid)));
                }
            }
        }
        match self.underlying {
            Some(underlying) => underlying.get(host, folder),
            None => Ok((None, None)),
        }
    }

    fn set(
        &self,
        host: &str,
        folder: &str,
        uid_validity: u32,
        last_uid: u32,
    ) -> Result<(), String> {
        self.staged.lock().expect("hwm store lock").push((
            host.to_string(),
            folder.to_string(),
            uid_validity,
            last_uid,
        ));
        Ok(())
    }
}
